tools/wptrunner/wptrunner/executors/base.py

```diff
@@ -34,7 +34,7 @@ def executor_kwargs(test_type, test_environment, run_info_data, subsuite, **kwar
                        "target_platform": run_info_data["os"]}
 
     if test_type in ("reftest", "print-reftest"):
-        executor_kwargs["screenshot_cache"] = test_environment.cache_manager.dict()
+        executor_kwargs["screenshot_cache"] = test_environment.screenshot_caches[test_type]
```

Comment:
Why this is needed?

Reply:
The [same reference](https://source.chromium.org/search?q=f:external%2Fwpt.*-print.html%20link.*rel%3D.*blank.html&sq=) can be used by both test types, so the screenshots might be different under pagination.

The reftest executor can't be responsible for clearing the cache because the cache should only be cleared after all tests run (to maximize hits), and each executor only lives for as long as it doesn't need to be restarted. The `TestEnvironment` exposing `screenshot_caches` lets it do that without needing to reach into the executor kwargs, which are encapsulated in `TestImplementation`.

Reply:
> The same reference can be used by both test types, so the screenshots might be different under pagination.

So we had this problem in the past but never noticed? I see [`viewport_size` is part of the key to access the cache](https://source.chromium.org/chromium/chromium/src/+/main:third_party/wpt_tools/wpt/tools/wptrunner/wptrunner/executors/base.py;l=420), and print-reftests do have a different viewport size  [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/wpt_tools/wpt/tools/wptrunner/wptrunner/wpttest.py;l=711).

Or maybe using `viewport_size` as the key is confusing and we should stick to `(subsuite, test_type, test_url)`?

[This line](https://source.chromium.org/chromium/chromium/src/+/main:third_party/wpt_tools/wpt/tools/wptrunner/wptrunner/executors/base.py;l=612) does not include subsuite name in the key. Is that a bug?

Reply:
> So we had this problem in the past but never noticed?

I'm not sure what you mean. The existing code already [creates one cache][0] per [(test type, subsuite) pair][1] (previous commit only created per-test-type, which the latest commit fixes). The refactor keeps that, but the `TestEnvironment` has a handle to them now.

Since caches for different subsuites are already isolated, I think we can remove `subsuite` from the cache key. Then the line without `subsuite` in the key is correct as-is.

> Or maybe using viewport_size as the key is confusing and we should stick to (subsuite, test_type, test_url)?

I would say just keep `viewport_size` for now, since the cache key should represent everything that could affect the screenshot. Some vendors may have nontrivial values.

[0]: https://source.chromium.org/chromium/chromium/src/+/main:third_party/wpt_tools/wpt/tools/wptrunner/wptrunner/executors/base.py;l=37;drc=55f1d455165e6583a93810949c68e79182269d58;bpv=0;bpt=0
[1]: https://source.chromium.org/chromium/chromium/src/+/main:third_party/wpt_tools/wpt/tools/wptrunner/wptrunner/wptrunner.py;l=237-242;drc=1aec6f9b305344c52fecb7a7b599e14ad2fbc88d;bpv=0;bpt=0

Reply:
SG